### ISSUE TITLE
`read_parquet_functions.py` not working when upgrading `pyarrow` to 12.0.1

### ISSUE THREAD

**Issue by lindamegner** [innosat-mats/MATS-L1-processing#149]
# 🐛 Description

`read_parquet_functions.py` crashes when I upgrade pyarrow from 10.0.1 to 12.0.0

# 🔢 To reproduce upgrade pyarrow and run this code snippet
```python
starttime=DT.datetime(2023,4,10,10,30,0)
endtime=DT.datetime(2023,4,10,11,0,0)

filter={'CCDSEL': [1, 4]} 
#filter={'CCDSEL': [1, 1], 'NRBIN': [1, 1], 'NCBINCCDColumns': [1, 1], 'NCOL':[2047,2048], 'NROW':[511,512]} 
#dfl1b = read_MATS_data(starttime, endtime,filter,level='1b',version='0.4')
dfl1a = read_MATS_data(starttime, endtime,filter, level='1a',version='0.5')
```

# ✌🏽 Expected behaviour

Should work.

# 💻 Traceback

```
Error messsage:

---------------------------------------------------------------------------
ArrowNotImplementedError                  Traceback (most recent call last)
[/Users/lindamegner/MATS/MATS-retrieval/MATS-analysis/Linda/flatfield_from_averaging.py](https://file+.vscode-resource.vscode-cdn.net/Users/lindamegner/MATS/MATS-retrieval/MATS-analysis/Linda/flatfield_from_averaging.py) in line 9
      [83](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-analysis/Linda/flatfield_from_averaging.py?line=82) filter={'CCDSEL': [1, 4]} 
      [84](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-analysis/Linda/flatfield_from_averaging.py?line=83) #filter={'CCDSEL': [1, 1], 'NRBIN': [1, 1], 'NCBINCCDColumns': [1, 1], 'NCOL':[2047,2048], 'NROW':[511,512]} 
      [85](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-analysis/Linda/flatfield_from_averaging.py?line=84) #dfl1b = read_MATS_data(starttime, endtime,filter,level='1b',version='0.4')
----> [86](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-analysis/Linda/flatfield_from_averaging.py?line=85) dfl1a = read_MATS_data(starttime, endtime,level='1a',version='0.5')

File [~/MATS/MATS-retrieval/MATS-utility-functions/src/mats_utils/rawdata/read_data.py:35](https://file+.vscode-resource.vscode-cdn.net/Users/lindamegner/MATS/MATS-retrieval/~/MATS/MATS-retrieval/MATS-utility-functions/src/mats_utils/rawdata/read_data.py:35), in read_MATS_data(start_date, end_date, filter, version, level)
     [32](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-utility-functions/src/mats_utils/rawdata/read_data.py?line=31) else:
     [33](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-utility-functions/src/mats_utils/rawdata/read_data.py?line=32)     filesystem = f"ops-payload-level{level}-v{version}"
---> [35](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-utility-functions/src/mats_utils/rawdata/read_data.py?line=34) ccd_data = read_ccd_data_in_interval(start_date, end_date, filesystem, s3,filter=filter)
     [37](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-utility-functions/src/mats_utils/rawdata/read_data.py?line=36) if level == '1a':
     [38](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-utility-functions/src/mats_utils/rawdata/read_data.py?line=37)     add_ccd_item_attributes(ccd_data)

File [~/MATS/MATS-retrieval/MATS-L1-processing/src/mats_l1_processing/read_parquet_functions.py:273](https://file+.vscode-resource.vscode-cdn.net/Users/lindamegner/MATS/MATS-retrieval/~/MATS/MATS-retrieval/MATS-L1-processing/src/mats_l1_processing/read_parquet_functions.py:273), in read_ccd_data_in_interval(start, stop, path, filesystem, filter, metadata)
    [267](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-L1-processing/src/mats_l1_processing/read_parquet_functions.py?line=266)     for variable in filter.keys():
    [268](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-L1-processing/src/mats_l1_processing/read_parquet_functions.py?line=267)         filterlist &= (
    [269](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-L1-processing/src/mats_l1_processing/read_parquet_functions.py?line=268)             (ds.field(variable) >= filter[variable][0])
    [270](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-L1-processing/src/mats_l1_processing/read_parquet_functions.py?line=269)             & (ds.field(variable) <= filter[variable][1])
    [271](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-L1-processing/src/mats_l1_processing/read_parquet_functions.py?line=270)         )
--> [273](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-L1-processing/src/mats_l1_processing/read_parquet_functions.py?line=272) table = dataset.to_table(filter=partition_filter & filterlist)
    [275](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-L1-processing/src/mats_l1_processing/read_parquet_functions.py?line=274) dataframe = table.to_pandas()
    [276](file:///Users/lindamegner/MATS/MATS-retrieval/MATS-L1-processing/src/mats_l1_processing/read_parquet_functions.py?line=275) dataframe.reset_index(inplace=True)

File [~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/_dataset.pyx:537](https://file+.vscode-resource.vscode-cdn.net/Users/lindamegner/MATS/MATS-retrieval/~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/_dataset.pyx:537), in pyarrow._dataset.Dataset.to_table()

File [~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/_dataset.pyx:383](https://file+.vscode-resource.vscode-cdn.net/Users/lindamegner/MATS/MATS-retrieval/~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/_dataset.pyx:383), in pyarrow._dataset.Dataset.scanner()

File [~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/_dataset.pyx:3202](https://file+.vscode-resource.vscode-cdn.net/Users/lindamegner/MATS/MATS-retrieval/~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/_dataset.pyx:3202), in pyarrow._dataset.Scanner.from_dataset()

File [~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/_dataset.pyx:3120](https://file+.vscode-resource.vscode-cdn.net/Users/lindamegner/MATS/MATS-retrieval/~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/_dataset.pyx:3120), in pyarrow._dataset.Scanner._make_scan_options()

File [~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/_dataset.pyx:3054](https://file+.vscode-resource.vscode-cdn.net/Users/lindamegner/MATS/MATS-retrieval/~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/_dataset.pyx:3054), in pyarrow._dataset._populate_builder()

File [~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/_compute.pyx:2558](https://file+.vscode-resource.vscode-cdn.net/Users/lindamegner/MATS/MATS-retrieval/~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/_compute.pyx:2558), in pyarrow._compute._bind()

File [~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/error.pxi:144](https://file+.vscode-resource.vscode-cdn.net/Users/lindamegner/MATS/MATS-retrieval/~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/error.pxi:144), in pyarrow.lib.pyarrow_internal_check_status()

File [~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/error.pxi:121](https://file+.vscode-resource.vscode-cdn.net/Users/lindamegner/MATS/MATS-retrieval/~/miniconda3/envs/mats-analysis/lib/python3.10/site-packages/pyarrow/error.pxi:121), in pyarrow.lib.check_status()

ArrowNotImplementedError: Function 'greater_equal' has no kernel matching input types (timestamp[ns, tz=UTC], timestamp[s])
```

**Comment by skymandr** [comment]
The problem is that the data read from the database has a timezone. I'm not sure if this is intentional or not, but I suspect you can get around the problem by setting a timezone on your dates:
```python
starttime = DT.datetime(2023, 4,10,10, 30, 0, tzinfo=DT.timezone.utc)
endtime = DT.datetime(2023, 4,10,11, 0, 0, tzinfo=DT.timezone.utc)
```

**Comment by skymandr** [comment]
@lindamegner is this resolved? If so, please close the issue.

**Comment by lindamegner** [comment]
I dont understand. Do you mean that if I upgrade pyarrow to a more recent version I always must set tzinfo=DT.timezone.utc as an extra variable? If so, is it not possible to add something that sets that varibale inside the code so that we don't have to add it in every script that calls it?

**Comment by skymandr** [comment]
I think that is a question for @OleMartinChristensen rather than me, since he maintains that code.

**Comment by OleMartinChristensen** [comment]
Tough I dont understand the headline of this issue, this problem has been adressed in read_MATS_data lines 23-26

```
    if start_date.tzinfo == None:
        start_date = start_date.replace(tzinfo=timezone.utc)
    if end_date.tzinfo == None:
        end_date = end_date.replace(tzinfo=timezone.utc)
```

**Comment by lindamegner** [comment]
I just tested to upgrade pyarrow to version 12.0.1 again, and it is still crashing with the error above. So that fix does not work.

**Comment by OleMartinChristensen** [comment]
ok, Ill have a look when Im back from holiday

**Comment by lindamegner** [comment]
This is still a problem.

**Comment by lindamegner** [comment]
@OleMartinChristensen

**Comment by skymandr** [comment]
Note that the processing chain uses 10.0.1 across the board. Looking at the release notes for 11 and 12, I see some changes have been made regarding timestamps (and some additional changes in 13) which is the latest, but nothing which is the obvious course for this issue. See https://arrow.apache.org/blog/

It should be possible to work around. One resource online mentions using `pa.scalar` in the filter, for instance, but maybe we should also consider upgrading arrow across the whole pipeline, if there are compelling reasons to do so.

**Comment by skymandr** [comment]
Revisiting this, @joakimmoller pointed out that the issue might not be with the timezones, but that one part in the comparison is in `[ns]` and the other in `[s]`.

**Comment by skymandr** [comment]
This is actually a bug in `pyarrow`, which I have verifed is fixed in `pyarrow>=14.0.0`, however, upgrading to that (or the latest stable, which is `pyarrow==15.0.0`), breaks some other things with filters. I will have a look and see if I can resolve that, but we should all avoid versions 11 to 13 it seems.
